Compiler/src/Compiler.cpp

```diff
@@ -1477,6 +1480,70 @@ struct Compiler
         }
     }
 
+    void compileExprInterpString(AstExprInterpString* expr, uint8_t target, bool targetTemp)
+    {
+        std::string formatString;
+
+        size_t stringsLeft = expr->strings.size;
+
+        for (AstArray<char> const& string : expr->strings)
+        {
+            std::string stringEscaped(string.data);
+
+            for (size_t characterIndex = 0; characterIndex < stringEscaped.size(); ++characterIndex)
+            {
+                if (stringEscaped[characterIndex] == '%')
+                {
+                    stringEscaped.insert(characterIndex, 1, '%');
+                    characterIndex++;
+                }
+            }
+
+            formatString += stringEscaped;
+
+            stringsLeft--;
+
+            if (stringsLeft > 0)
+                formatString += "%*";
+        }
+
+        auto formatStringSize = formatString.size();
+
+        // We can't use formatStringRef.data() directly, because short strings don't have their data
+        // pinned in memory, so when interpFormatStrings grows, these pointers will move and become invalid.
+        std::shared_ptr<char[]> formatStringPtr(new char[formatStringSize]);
+        memcpy(formatStringPtr.get(), formatString.data(), formatStringSize);
+
+        auto formatStringPtrRef = interpFormatStrings.emplace_back(formatStringPtr);
+        AstArray<char> formatStringArray{formatStringPtrRef.get(), formatStringSize};
+
+        int32_t formatStringIndex = bytecode.addConstantString(sref(formatStringArray));
+        if (formatStringIndex < 0)
+            CompileError::raise(expr->location, "Exceeded constant limit; simplify the code to compile");
+
+        // INTERP CODE REVIEW: Why do I need this?
```

Comment:
The register structure below is incorrect. I believe you can reproduce the bug by using code like this:

```
local a, b = 1, c = 2

a = `foo{42}`
print(a, b, c) -- expecting foo42 1 2, but you'll actually get foo42 nil nil I believe
```

You can't compile a CALL of any form that targets arbitrary register. Instead you need to allocate the registers for the call frame - I think you need 2+expr->expressions.size here - and use the resulting registers to execute a CALL, and then at the end MOVE the result back to target. You also need RegScope to make sure the registers are freed at the end of the call.

Reply:
Even weirder:

```lua
local a, b, c = nil, "um", "uh oh"
a = `foo{"bar"}`
print(a) -- foouh oh
```

Reply:
I'm not sure on the specifics here. Can you provide more detail on how this would look in code?

Reply:
Something like (untested):

```
RegScope rs(this);

uint8_t regs = allocReg(expr, 2 + expr->expressions.size);

emitLoadK(regs, formatStringConstantIndex);
for (size_t i = 0; i < expr->expressions.size; ++i)
  compileExprTempTop(expr->expressions.data[i], uint8_t(regs + 2 + i));

bytecode.emitABC(LOP_NAMECALL, regs, regs, getStringHash(format));
bytecode.emitAux(formatIndex);
bytecode.emitABC(LOP_CALL, regs, uint8_t(1 + expr->expressions.size + 1), 1 + 1);
bytecode.emitABC(LOP_MOVE, target, regs, 0);
```

Reply:
With a bit of finagling, I can get that to work, but I cannot seem to figure out how to optimize the LOP_MOVE out when it is unnecessary, which standard namecalls can accomplish, and I do not want to have branches where the interp string format is less performant.